### PR TITLE
Multiple database instances

### PR DESCRIPTION
**Pull request by ybelMekk**
This PR, will - not pretty - give users the ability to specify several of databases and users for each instance. Several instances is switched off for now with an if, like before. It could possibly be the best solution for instances if instance.Name was required from the beginning, as it is for database.Name.

This PR solves the issue - not in a pretty way - and it also ads some complexity. 

Only things I can come and think about that will get changed  of resources is if an application has specified:
`databases.Users[]`, with specified users, will get new secrets and new sqlusers.

All defaults will still be as before. So no default users will be affected. 

Also i removed the ability to auto generate a default user for the 2 or 3 etc. database, think this should been mandatory field to specify a database user.Name anyway.. makes things and life so much easier. 

good weekend

### CODE REVIEW

File: pkg/resourcecreator/google/sql/instance.go [hunk 1]
```diff
@@ -73,15 +82,26 @@ func GoogleSqlInstance(objectMeta metav1.ObjectMeta, instance nais.CloudSqlInsta
...
+	suffix := fmt.Sprintf("%s-%s", resourceType, number)
+	basename := fmt.Sprintf("%s-%s", defaultReturn, suffix)
+	return BuildUniquesNameWithPredicate(number == "0", defaultReturn, basename)
+}
+
+func CloudSqlInstanceWithDefaults(instance nais.CloudSqlInstance, appName string, instanceNumber int) (nais.CloudSqlInstance, error) {
 	var err error
 
+	// Due to the fact that instance.Name is not required, unique name for instances need to be generated
+	instanceName, err := UniqueName(instanceNumber, appName, "instance")
```
Comment (Kyrremann) [inline pkg/resourcecreator/google/sql/instance.go:96]: Mangler `err`-sjekk. Samme på linje 98.

File: pkg/resourcecreator/google/sql/instance.go [hunk 2]
```diff
@@ -73,15 +82,26 @@ func GoogleSqlInstance(objectMeta metav1.ObjectMeta, instance nais.CloudSqlInsta
...
+	// This default will always be overridden by GoogleSQLDatabase(), but need to be set her as database.Name can not be nil.
+	dbName, err := UniqueName(instanceNumber, appName, "db")
```
Comment (Kyrremann) [inline pkg/resourcecreator/google/sql/instance.go:98]: Hvis `dbName` blir overskrevet senere, er det ikke bedre med `dummy-name` eller noe i den dur? Slik at det er tydelig at dette er et midlertidig navn som blir overskrevet.

File: pkg/resourcecreator/google/sql/instance_test.go [hunk 3]
```diff
@@ -47,12 +47,26 @@ func TestGoogleSqlInstance(t *testing.T) {
...
+
+		speca := nais.CloudSqlInstance{
+			Name: appa.Name,
+			Type: nais.CloudSqlInstanceTypePostgres12,
+		}
+
+		speca, err = google_sql.CloudSqlInstanceWithDefaults(speca, appa.Name, 1)
+		assert.NoError(t, err)
+		assert.Equal(t, "myapplication-instance-1-36663990", speca.Name)
+		// Not really necessary to test, this will be overridden any day buy required Name in database field.
```
Comment (Kyrremann) [inline pkg/resourcecreator/google/sql/instance_test.go:69]: Skjønner ikke helt hva du mener med denne kommentaren. Er det slik at navnet snart blir overskrever (any day now?), eller at dette er et felt som brukeren setter?

File: pkg/resourcecreator/google/sql/user_test.go [hunk 4]
```diff
@@ -189,9 +190,14 @@ func TestMergeDefaultSQLUser(t *testing.T) {
...
+	assert.NoError(t, err)
 	assert.Equal(t, []nais.CloudSqlDatabaseUser{{Name: instance.Name}}, mergedUsers)
 
-	mergedUsers = google_sql.MergeAndFilterSQLUsers(dbUsers, instance.Name)
+	mergedUsers, err = google_sql.MergeAndFilterDatabaseSQLUsers(dbUsers, instance.Name, 0)
+	assert.NoError(t, err)
 	assert.Equal(t, expected, mergedUsers)
+
+	mergedUsers, err = google_sql.MergeAndFilterDatabaseSQLUsers(nil, instance.Name, 1)
+	assert.Error(t, err, fmt.Errorf("must to specify users for extra databases, can not have several databases with default user"))
```
Comment (Kyrremann) [inline pkg/resourcecreator/google/sql/user_test.go:202]:
```suggestion
	assert.Error(t, err, fmt.Errorf("must specify users for each extra databases, can not have several databases with the default user"))
```

File: pkg/resourcecreator/testdata/gcp_database_serveral_databases_users.yaml [hunk 5]
```diff
@@ -0,0 +1,350 @@
+config:
```
Comment (Kyrremann) [inline pkg/resourcecreator/testdata/gcp_database_serveral_databases_users.yaml:1]: Filen har skrivefeil i navnet, skal være `several`, ikke `serveral`.

File: pkg/resourcecreator/testdata/naisjob/cronjob_gcp_database_serveral_databases_users.yaml [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+                - name: User_two
+                - name: _user_three3
+            - name: mydb2
+              users:
+                - name: User_four
+                - name: _user_five
+          type: POSTGRES_11
+
+tests:
+  - match:
```
Comment (Kyrremann) [inline pkg/resourcecreator/testdata/naisjob/cronjob_gcp_database_serveral_databases_users.yaml:35]: Trenger man virkelig denne testen i så mange testdata-filer?

**Review by Kyrremann — COMMENTED** [review]
Syns dette ser veldig bra ut! Og virker som det skal være bakoverkompatibelt!

File: pkg/resourcecreator/google/sql/instance_test.go (hunk 3 above)
Comment (ybelMekk) [reply pkg/resourcecreator/google/sql/instance_test.go:69]: den vil alltid bli overskriver pga. databases.Name er reuired, men fjern testen

File: pkg/resourcecreator/google/sql/instance.go (hunk 2 above)
Comment (ybelMekk) [reply pkg/resourcecreator/google/sql/instance.go:98]: done.

File: pkg/resourcecreator/testdata/gcp_database_serveral_databases_users.yaml (hunk 5 above)
Comment (ybelMekk) [reply pkg/resourcecreator/testdata/gcp_database_serveral_databases_users.yaml:1]: aaa..

File: pkg/resourcecreator/testdata/naisjob/cronjob_gcp_database_serveral_databases_users.yaml (hunk 6 above)
Comment (ybelMekk) [reply pkg/resourcecreator/testdata/naisjob/cronjob_gcp_database_serveral_databases_users.yaml]: er ikke sikker på hva du mener her?

**Comment by ybelMekk** [comment]
[Støtte for flere databaser i Naiserator](https://trello.com/c/hZ6Ebtvv/70-st%C3%B8tte-for-flere-databaser-i-naiserator)

**Comment by ybelMekk** [comment]
Continue this when evaluation of survey is done.

**Pull request closed without merging**